ambari-server/src/main/java/org/apache/ambari/server/stack/NameService.java

```diff
@@ -0,0 +1,153 @@
+/*
+ *
+ *  * Licensed to the Apache Software Foundation (ASF) under one
+ *  * or more contributor license agreements.  See the NOTICE file
+ *  * distributed with this work for additional information
+ *  * regarding copyright ownership.  The ASF licenses this file
+ *  * to you under the Apache License, Version 2.0 (the
+ *  * "License"); you may not use this file except in compliance
+ *  * with the License.  You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.apache.ambari.server.stack;
+
+import static org.apache.commons.lang.StringUtils.isBlank;
+
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.apache.ambari.server.state.Cluster;
+import org.apache.ambari.server.state.ConfigHelper;
+import org.apache.ambari.server.utils.HTTPUtils;
+import org.apache.ambari.server.utils.HostAndPort;
+import org.apache.commons.lang.builder.ToStringBuilder;
+
+public class NameService {
+
+  public static class NameNode {
+    private final String uniqueId;
+    private final String address;
+    private final boolean encrypted;
+    private final String propertyName;
+
+    public static NameNode fromConfig(String nameServiceId, String nnUniqueId, ConfigHelper config, Cluster cluster) {
```

Comment:
I feel like having a class for a specific component of a given stack is a bad practice. Can this be made into some metainfo that ships with the stack so it's not hard coded?

Reply:
I don't think writing ambari in a fully component or stack agnostic way is possible (we already have lots of switch cases on service or component names). This feature is about handling a component specific setup. We need to know if the namenode is in federated mode and we only know this based on a component specific config file. This class represents a real domain concept and it just makes easier to extract this information from the config. Do you have a concrete idea how could we extract this info in a more dynamic way?

Reply:
Alerts did something similar with HA modes where it was driven by stack properties, like this:

`            "high_availability": {
              "nameservice": "{{hdfs-site/dfs.internal.nameservices}}",
              "alias_key" : "{{hdfs-site/dfs.ha.namenodes.{{ha-nameservice}}}}",
              "http_pattern" : "{{hdfs-site/dfs.namenode.http-address.{{ha-nameservice}}.{{alias}}}}",
              "https_pattern" : "{{hdfs-site/dfs.namenode.https-address.{{ha-nameservice}}.{{alias}}}}"
            }`

If we added this information to the HDFS metainfo.xml for NameNode, then you wouldn't need to hardcode it in here. You could instead make this a more generic class that could work for other pairs, like YARN's RM.

Reply:
This makes sense but I think this is out of scope now. Anyways most of these properties were already there, I've just modified the way we parse one of them (in federation mode one can contain comma separated ids) and moved them to a separate class. How about making a jira about this improvement?

Reply:
Ok - let's do that. I'm totally fine following up on this with a 2nd Jira.